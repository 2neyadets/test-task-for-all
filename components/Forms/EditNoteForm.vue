<template lang="pug">
  form.edit-note-form(v-if="note" @submit.prevent="saveNote")
    .edit-note-form__editable
      MyInput.pb-lg(
        :value="note.title"
        @input="v => handleChange(note, 'title', v)"
        label="Название заметки"
        :maxlength="20"
        clearable
      )
      template(v-for="(task, index) in note.tasks")
        .task.row.flex-center
          MyCheckbox.mb-lg(
            @inputDone="v => handleChange(task, 'done', v)"
            @inputText="v => handleChange(task, 'text', v)"
            :task="task"
            :key="task.id + index"
            label="Текст задачи"
            editable
            clearable
          )
          MyBtn.task__delete-btn.ml-sm.mb-lg(
            @click.native="deleteTask(task.id)"
            icon="delete"
            size="26"
          )
      .flex.flex-center(v-if="isDisableCreateNewTask")
        p Максимальное количество задач для одной заметки - 5
      .flex.flex-center
        MyBtn.add-task-btn(
          @click.native="addTask"
          label="Добавить задачу"
          icon="add"
          :disable="isDisableCreateNewTask"
          autofocus
        )
    .edit-note-form__actions
      .edit-note-form__actions--normal
        .flex.flex-center
          MyBtn.save-changes-btn(
            type="submit"
            label="Сохранить изменения"
            icon="save"
            :disable="!isNoteChanged || isNoteHaveEmptyTextFields"
          )
        .flex.flex-center.pt-md
          MyBtn.undo-btn(
            @click.native="undoLastChange"
            label="Отменить внесенное изменение"
            icon="clear"
            :disable="!isAllow.undo"
          )
        .flex.flex-center.pt-md
          MyBtn.redo-btn(
            @click.native="redoLastChange"
            label="Вернуть отменённое изменение"
            icon="repeat"
            :disable="!isAllow.redo"
          )
      .edit-note-form__actions--danger
        .flex.flex-center
          MyBtn.cancel-edit-btn(
            @click.native="openCancelEditNoteDialog"
            label="Отменить редактирование"
            icon="cancel"
          )
        .flex.flex-center.pt-md
          MyBtn.delete-note-btn(
            @click.native="openDeleteDialog"
            label="Удалить заметку"
            icon="delete"
          )
</template>

<script>
import MyInput from '../Reusable/MyInput'
import MyBtn from '../Reusable/MyBtn'
import MyCheckbox from '../Reusable/MyCheckbox'
import { getId } from '../../assets/helpers'

export default {
  name: 'EditNoteForm',
  components: { MyCheckbox, MyBtn, MyInput },
  props: {
    currentNote: {
      type: Object,
      required: true,
    }
  },
  data () {
    return {
      lastNote: null,
      note: null,
      isAllow: {
        undo: false,
        redo: false,
      },
    }
  },
  computed: {
    isNoteChanged () {
      return JSON.stringify(this.note) !== JSON.stringify(this.currentNote)
    },
    isNoteHaveEmptyTextFields () {
      return !this.note.title || (this.note.tasks && this.note.tasks.some(task => (!task.text)))
    },
    isDisableCreateNewTask () {
      return this.note.tasks && this.note.tasks.length >= 5
    },
  },
  watch: {
    currentNote (v) {
      this.initNotes()
    }
  },
  created () {
    this.initNotes()
  },
  mounted () {
  },
  updated () {},
  methods: {
    initNotes () {
      this.lastNote = JSON.parse(JSON.stringify(this.currentNote))
      this.note = JSON.parse(JSON.stringify(this.currentNote))
    },
    allowUndoAndWriteLastNote () {
      this.isAllow.undo = true
      this.isAllow.redo = false
      this.lastNote = JSON.parse(JSON.stringify(this.note))
    },
    swapNotes () {
      const temp = JSON.parse(JSON.stringify(this.note))
      this.note = JSON.parse(JSON.stringify(this.lastNote))
      this.lastNote = temp
    },
    handleChange (obj, key, v) {
      this.allowUndoAndWriteLastNote()
      obj[key] = v
    },
    addTask () {
      if (!this.note.tasks) this.note = { ...this.note, tasks: [] }
      if (!this.isDisableCreateNewTask) {
        this.note.tasks.push({
          id: getId(),
          done: false,
          text: null,
        })
      }
    },
    deleteTask (taskId) {
      this.allowUndoAndWriteLastNote()
      this.note.tasks.splice(this.note.tasks.findIndex(task => task.id === taskId), 1)
    },
    undoLastChange () {
      this.swapNotes()
      this.isAllow.undo = false
      this.isAllow.redo = true
    },
    redoLastChange () {
      this.swapNotes()
      this.isAllow.undo = true
      this.isAllow.redo = false
    },
    openCancelEditNoteDialog () {
      this.$store.dispatch('changeDialogState', 'cancelEditNote')
    },
    openDeleteDialog () {
      this.$store.dispatch('changeDialogState', 'deleteNote')
    },
    saveNote () {
      if (!this.isNoteChanged || this.isNoteHaveEmptyTextFields) return false
      this.$store.dispatch('notes/updateNoteAction', JSON.parse(JSON.stringify(this.note)))
    },
  },
}
</script>

<style scoped lang="stylus">
  .edit-note-form

    &__editable
      border 1px solid white
      padding 8px
    &__actions

      &--normal
        border 1px solid green
        padding 8px
        margin-top 8px

      &--danger
        border 1px solid red
        padding 8px
        margin-top 8px
</style>
