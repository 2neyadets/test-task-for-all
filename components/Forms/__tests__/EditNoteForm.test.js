import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import EditNoteForm from '../EditNoteForm'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('EditNoteForm component', () => {
  let wrapper, actions, noteGetters, store
  const currentNote = {
    id: 'f1733253812c-f300272d',
    title: 'Название заметки',
  }

  beforeEach(() => {
    actions = {
      'notes/updateNoteAction': jest.fn(),
      changeDialogState: jest.fn(),
    }
    noteGetters = {
      notesArr: () => [],
    }
    store = new Vuex.Store({
      actions,
      modules: {
        notes: {
          namespaced: true,
          getters: noteGetters,
        },
      },
    })

    wrapper = shallowMount(EditNoteForm, {
      localVue,
      store,
      propsData: {
        currentNote,
      },
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick, если нужно что-то выпонять внутри него
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('cant create task when have 5 tasks', async () => {
    expect.assertions(1)
    wrapper.setData({
      note: { ...currentNote, tasks: [{}, {}, {}, {}, {}], }
    })
    await localVue.nextTick()
    wrapper.vm.addTask()
    expect(wrapper.vm.$data.note.tasks.length).toBe(5)
  })

  it('can create task', async () => {
    expect.assertions(2)
    const btn = wrapper.find('.add-task-btn')
    btn.trigger('click')
    expect(wrapper.vm.$data.note.tasks.length).toBe(1)
    await localVue.nextTick()
    expect(wrapper.find('.task').exists()).toBe(true)
  })

  it('can delete task', async () => {
    expect.assertions(2)
    wrapper.setData({
      note: {
        id: 'f1733253812c-f300272d',
        title: 'Название заметки',
        tasks: [
          {
            done: false,
            id: 'f173325392ff-f4a0c9a4',
            text: 'Текст задачи',
          }
        ],
      }
    })
    await localVue.nextTick()
    expect(wrapper.find('.task').exists()).toBe(true)
    const btn = wrapper.find('.task__delete-btn')
    btn.trigger('click')
    await localVue.nextTick()
    expect(wrapper.find('.task').exists()).toBe(false)
  })

  it('can save changes', async () => {
    expect.assertions(1)
    wrapper.setData({
      note: {
        id: 'f1733253812c-f300272d',
        title: 'Новое название',
      }
    })
    await localVue.nextTick()
    const btn = wrapper.find('.save-changes-btn')
    btn.trigger('submit')
    expect(actions['notes/updateNoteAction']).toHaveBeenCalled()
  })

  it('can undo change', async () => {
    expect.assertions(1)
    wrapper.setData({
      note: {
        id: 'f1733253812c-f300272d',
        title: 'asdasdsadasdasdasd',
      }
    })
    await localVue.nextTick()
    const btn = wrapper.find('.undo-btn')
    btn.trigger('click')
    await localVue.nextTick()
    expect(wrapper.vm.$data.note).toEqual(currentNote)
  })

  it('can redo change after undo', async () => {
    expect.assertions(7)
    const firstChange = 'first',
      secondChange = 'second',
      myInput = wrapper.find('MyInput-stub'),
      undoBtn = wrapper.find('.undo-btn'),
      redoBtn = wrapper.find('.redo-btn')
    myInput.vm.$emit('input', firstChange)
    expect(myInput.exists()).toBe(true)
    expect(undoBtn.exists()).toBe(true)
    expect(redoBtn.exists()).toBe(true)
    await localVue.nextTick()
    expect(wrapper.vm.$data.note.title).toBe(firstChange)
    myInput.vm.$emit('input', secondChange)
    await localVue.nextTick()
    expect(wrapper.vm.$data.note.title).toBe(secondChange)
    undoBtn.trigger('click')
    await localVue.nextTick()
    expect(wrapper.vm.$data.note.title).toBe(firstChange)
    redoBtn.trigger('click')
    await localVue.nextTick()
    expect(wrapper.vm.$data.note.title).toBe(secondChange)
  })

  it('watcher on currentNote works', async () => {
    expect.assertions(2)
    const newNote = { ...currentNote, title: 'новое название' }
    wrapper.setProps({
      currentNote: newNote,
    })
    await localVue.nextTick()
    expect(wrapper.vm.$data.note).toEqual(newNote)
    expect(wrapper.vm.$data.lastNote).toEqual(newNote)
  })

  it('can open "cancel edit" dialog', () => {
    expect.assertions(1)
    const btn = wrapper.find('.cancel-edit-btn')
    btn.trigger('click')
    expect(actions.changeDialogState).toHaveBeenCalled()
  })

  it('can open "delete note" dialog', () => {
    expect.assertions(1)
    const btn = wrapper.find('.delete-note-btn')
    btn.trigger('click')
    expect(actions.changeDialogState).toHaveBeenCalled()
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
