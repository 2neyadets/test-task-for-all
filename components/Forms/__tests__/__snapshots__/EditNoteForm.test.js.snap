// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNoteForm component match snapshot 1`] = `
<form
  class="edit-note-form"
>
  <div
    class="edit-note-form__editable"
  >
    <myinput-stub
      class="pb-lg"
      clearable="true"
      hint=""
      label="Название заметки"
      maxlength="20"
      mockid=""
      value="Название заметки"
    />
    <!---->
    <div
      class="flex flex-center"
    >
      <mybtn-stub
        autofocus="true"
        class="add-task-btn"
        icon="add"
        label="Добавить задачу"
        size="20"
        type="button"
      />
    </div>
  </div>
  <div
    class="edit-note-form__actions"
  >
    <div
      class="edit-note-form__actions--normal"
    >
      <div
        class="flex flex-center"
      >
        <mybtn-stub
          class="save-changes-btn"
          disable="true"
          icon="save"
          label="Сохранить изменения"
          size="20"
          type="submit"
        />
      </div>
      <div
        class="flex flex-center pt-md"
      >
        <mybtn-stub
          class="undo-btn"
          disable="true"
          icon="clear"
          label="Отменить внесенное изменение"
          size="20"
          type="button"
        />
      </div>
      <div
        class="flex flex-center pt-md"
      >
        <mybtn-stub
          class="redo-btn"
          disable="true"
          icon="repeat"
          label="Вернуть отменённое изменение"
          size="20"
          type="button"
        />
      </div>
    </div>
    <div
      class="edit-note-form__actions--danger"
    >
      <div
        class="flex flex-center"
      >
        <mybtn-stub
          class="cancel-edit-btn"
          icon="cancel"
          label="Отменить редактирование"
          size="20"
          type="button"
        />
      </div>
      <div
        class="flex flex-center pt-md"
      >
        <mybtn-stub
          class="delete-note-btn"
          icon="delete"
          label="Удалить заметку"
          size="20"
          type="button"
        />
      </div>
    </div>
  </div>
</form>
`;
